### ISSUE TITLE
Websocket Provider constantly disconneting / reconnecting, resulting in BAN by Infura

### ISSUE THREAD

**Issue by bizso09** [ethereum/web3.py#1770]
In this code:

```
    async def __aenter__(self) -> websockets.WebSocketClientProtocol:
        if self.ws is None:
            self.ws = await websockets.connect(
                uri=self.endpoint_uri, loop=self.loop, **self.websocket_kwargs
            )
        return self.ws

    async def __aexit__(
        self, exc_type: Type[BaseException], exc_val: BaseException, exc_tb: TracebackType
    ) -> None:
        if exc_val is not None:
            try:
                await self.ws.close()
            except Exception:
                pass
            self.ws = None
```

For each request, a new websocket is opened, and once the request finishes, it is closed. When running multiple connections in parallel, this results in a BAN by Infura.io , because of the constant reconnections. In addition, constant reconnection defeats the whole purpose of using a websocket, which is suitable for streaming communication.

The websocket should be opened once at the initialization of the Websocket provider, and kept open through out its lifetime, not closed/reopened for each request individually. 

The fact that the class is called `PersistentWebsocket`, is also a misnomer, because the communication inside it is anything but persistent.

In addition, it is not clear to me if this supports asynchronous response handling, because all the async logic seems to be wrapped in a single thread. It would be nice if I could pass in callback functions, like in web3 js.

**Comment by kclowes** [comment]
From a brief glance at the code base, it does look like it connects with every request @bizso09. It will probably take us a while to get to this, but if you'd like to open a PR it would likely be faster and much appreciated!

> In addition, it is not clear to me if this supports asynchronous response handling

I'm pretty sure (but not 100%) that this does not support asynchronous response handling.

**Comment by NavyaZaveri** [comment]
I'd like to take a stab at this :) 

I think the easiest/quickest way to accomplish this is to make sure the socket  isn't closed/set to `None` on `__aexit__.` 

`I'm not sure *when* the socket should be closed though. I think a potential solution could be to introduce a custom destructor (`__del__`)  on `WebsocketProvider`, which would close the socket when it's out of scope

**Comment by kclowes** [comment]
Oh, you're right @NavyaZaveri. Thanks for digging in! I also noticed that it only re-connects if there is a system exception (anything that inherits from `BaseException`), rather than `Exception`, so won't disconnect on the "normal" errors that we'd get back from hitting JSON-RPC endpoints. I'm going to close, but if I missed something, feel free to reopen.
